synapse/neuron.py

```diff
@@ -446,91 +445,182 @@ def initConfDefs(self):
+                               'ephem_pub': self._my_ephem_prv.public().dump(),
+                               'cert': self._sess_boss.certbyts}))
 
-        skey = self._sess_boss.decrypt(data)
+    def _handle_session_msg(self, mesg):
+        ''' Validate and set up the crypto from a helo message '''
+        if self._crypter:
+            raise s_exc.ProtoErr('Received two client helos')
 
-        self.setRxKey(skey)
+        if self.is_lisn:
+            self._my_ephem_prv = s_ecc.PriKey.generate()
 
-    def sendcert(self):
-        self.link.tx(('cert', {'cert': self._sess_boss.certbyts}))
+        version = mesg[1].get('version')
+        if version != NEURON_PROTO_VERSION:
+            raise s_exc.ProtoErr('Found peer with missing or incompatible version')
 
-    @s_glob.inpool
-    def _onMesgCert(self, link, mesg):
+        peer_cert = s_vault.Cert.load(mesg[1].get('cert'))
+        peer_ephem_pub = s_ecc.PubKey.load(mesg[1].get('ephem_pub'))
 
-        if self.lisn:
-            self.sendcert()
+        if not self._sess_boss.valid(peer_cert):
+            clsn = self.__class__.__name__
+            raise s_exc.CryptoErr(mesg='%s got bad cert (%r)' % (clsn, peer_cert.iden(),))
 
-        cert = s_vault.Cert.load(mesg[1].get('cert'))
+        peer_static_pub = s_ecc.PubKey.load(peer_cert.tokn.get('ecdsa:pubkey'))
+        km = s_ecc.doECDHE(self._my_ephem_prv, peer_ephem_pub,
+                           self._sess_boss._my_static_prv, peer_static_pub, info=b'session')
 
-        if not self._sess_boss.valid(cert):
-            clsn = self.__class__.__name__
-            logger.warning('%s got bad cert (%r)' % (clsn, cert.iden(),))
-            self.fini()
+        to_initiator_symkey, to_listener_symkey = km[:32], km[32:]
+
+        if self.is_lisn:
+            self._crypter = CryptSeq(to_listener_symkey, to_initiator_symkey)
+        else:
+            self._crypter = CryptSeq(to_initiator_symkey, to_listener_symkey)
+            # Decrypt the first i.e. test message
+            first_msg_ct = mesg[1].get('first_msg')
+            self._crypter.decrypt(first_msg_ct)
+
+        return peer_cert
+
+    @s_glob.inpool
+    def _onMesgHelo(self, link, mesg):
+        '''
+        handle receiving the session establishment message from the peer.
+
+        send back our ephemerical public, our cert, and, if the listener, an encrypted message
+        '''
+
+        try:
+            peer_cert = self._handle_session_msg(mesg)
+        except Exception as e:
+            logger.exception('Exception encountered handling session message.')
+            self._send_fail(s_common.getexcfo(e))
+            self.txfini()
             return
 
-        # send back an skey message with our tx key
-        data = cert.public().encrypt(self.txkey)
-        self.link.tx(('skey', {'data': data}))
+        if self.is_lisn:
+            # This would be a good place to stick version or info stuff
+            first_message = {'first_message': 0}
```

Comment:
Could the first message just be to transmit a None instance to save us some wire bytes

Reply:
ok